src/Servers/HttpSys/src/RequestProcessing/Request.cs

```diff
@@ -351,6 +354,34 @@ private AspNetCore.HttpSys.Internal.SocketAddress LocalEndPoint
         }
     }
 
+    public ReadOnlyCollection<long> RequestTimestamps
+    {
+        get
+        {
+            if (_requestTimestamps == null)
+            {
+                /*
+                    Below is the definition of the timing info structure we are accessing the memory for.
+                    We can skip the first ULONG since it's always set to HttpRequestTimingTypeMax which is the size of the array.
+
+                    TODO: I would expect 244 bytes (4 bytes + 8 bytes * 30 [current size of HttpRequestTimingTypeMax]) but it's 248.
+                    It seems the RequestTimingCount is stored in 8 bytes as a ULONGLONG but I'm not sure why yet.
+
+                    typedef struct _HTTP_REQUEST_TIMING_INFO
+                    {
+                        ULONG RequestTimingCount;
+                        ULONGLONG RequestTiming[HttpRequestTimingTypeMax];
+
+                    } HTTP_REQUEST_TIMING_INFO, *PHTTP_REQUEST_TIMING_INFO;
+                */
+                var timings = RequestInfo[(int)HttpApiTypes.HTTP_REQUEST_INFO_TYPE.HttpRequestInfoTypeRequestTiming].Span.Slice(sizeof(long));
+
+                _requestTimestamps = new ReadOnlyCollection<long>(MemoryMarshal.Cast<byte, long>(timings).ToArray());
```

Comment:
While this does result in an allocation, I wasn't sure of a better way to expose the data as an IEnumerable since a ReadOnlySpan is a ref struct. Even if we remove this property, I did a quick benchmark between doing this vs doing the MemoryMarshal.Cast call every time a TryGet method is called and this was a bit more performant.

Reply:
Below is what I wrote for the benchmark. I tried to mimic someone calling TryGetTimestamp a various number of times. As you can see, the cost of calling ToArray averages out over the calls. The more calls made the better using ToArray is but even with just 6 calls it's close to identical. Let me know if you think my benchmark isn't correct or if you have other ideas to try. 

Also, per my comment about not using ReadOnlySpan, I think calling ToArray or something equivalent is required here to expose the Timestamps property but I could be wrong. 

> BenchmarkDotNet=v0.13.5, OS=Windows 10 (10.0.19045.3030/22H2/2022Update)
> Intel Core i7-7700 CPU 3.60GHz (Kaby Lake), 1 CPU, 8 logical and 4 physical cores
> .NET SDK=7.0.300-preview.23122.5
>   [Host]     : .NET 7.0.5 (7.0.523.17405), X64 RyuJIT AVX2
>   DefaultJob : .NET 7.0.5 (7.0.523.17405), X64 RyuJIT AVX2
> 
> |              Method | GetTimesatmpCallCount |      Mean |    Error |   StdDev |
> |-------------------- |---------------------- |----------:|---------:|---------:|
> | GetTimestampMarshal |                     1 |  20.12 ns | 0.455 ns | 0.576 ns |
> |  GetTimestampCached |                     1 |  36.04 ns | 0.543 ns | 0.482 ns |
> | GetTimestampMarshal |                     6 |  38.25 ns | 0.452 ns | 0.401 ns |
> |  GetTimestampCached |                     6 |  46.91 ns | 0.656 ns | 0.512 ns |
> | GetTimestampMarshal |                    18 |  87.13 ns | 1.196 ns | 1.060 ns |
> |  GetTimestampCached |                    18 |  64.07 ns | 0.725 ns | 0.605 ns |
> | GetTimestampMarshal |                    30 | 145.11 ns | 1.331 ns | 1.245 ns |
> |  GetTimestampCached |                    30 |  83.02 ns | 1.111 ns | 1.484 ns |


```csharp
    public class RequestTiming 
    {
        private readonly ReadOnlyMemory<byte> data;

        public RequestTiming()
        {
            var values = Enumerable.Range(1, 30).Select(i => long.MaxValue - (i * 100)).ToArray();
            data = new ReadOnlyMemory<byte>(MemoryMarshal.AsBytes(values.AsSpan()).ToArray());
        }

        [Params(1, 6, 18, 30)]
        public int GetTimesatmpCallCount { get; set; }

        [Benchmark]
        public List<long> GetTimestampMarshal()
        {
            var values = new List<long>(30);
            for (int i = 0; i < GetTimesatmpCallCount; i++)
            {
                values.Add(GetTimestamp(i));
            }

            return values;
        }

        [Benchmark]
        public List<long> GetTimestampCached()
        {
            long[] valuesArray = MemoryMarshal.Cast<byte, long>(data.Span).ToArray();
            var values = new List<long>(30);
            for (int i = 0; i < GetTimesatmpCallCount; i++)
            {
                values.Add(valuesArray[i]);
            }

            return values;
        }

        [MethodImpl(MethodImplOptions.NoInlining)]
        public long GetTimestamp(int index)
        {
            ReadOnlySpan<long> values = MemoryMarshal.Cast<byte, long>(data.Span);
            return values[index];
        }
    }
```

Reply:
I don't think it's fair to allocate the returned list in the Marshal case that does not allocate. I single call to `MemoryMarshal.Cast` is less than two nanoseconds if you don't penalize it for allocating list.

I don't see why we would go out of our way to disallow inlining either. Furthermore, we don't have to use `MemoryMarshal.Cast`. We can use `MemoryMarshal.CreateSpan` with some `Unsafe` code to avoid some of the checked arithmetic.

If we use `Unsafe`, even if you access all 30 fields per request recreating the span each time, it's still faster than `ToArray()` at 38 ns vs 45 ns.

> BenchmarkDotNet=v0.13.0, OS=Windows 10.0.22621
> Intel Core i9-10850K CPU 3.60GHz, 1 CPU, 20 logical and 10 physical cores
> .NET SDK=8.0.100-preview.5.23257.1
>   [Host]     : .NET 8.0.0 (8.0.23.26003), X64 RyuJIT
>   Job-WWRTZM : .NET 8.0.0 (8.0.23.25213), X64 RyuJIT
> 
> Server=True  Toolchain=.NET Core 8.0  RunStrategy=Throughput
> |              Method | GetTimestampCallCount |      Mean |     Error |    StdDev |          Op/s |  Gen 0 | Gen 1 | Gen 2 | Allocated |
> |-------------------- |---------------------- |----------:|----------:|----------:|--------------:|-------:|------:|------:|----------:|
> | GetTimestampMarshal |                     1 |  1.796 ns | 0.0271 ns | 0.0241 ns | 556,911,549.1 |      - |     - |     - |         - |
> |  GetTimestampCached |                     1 | 24.648 ns | 0.3853 ns | 0.3604 ns |  40,571,493.6 | 0.0010 |     - |     - |     264 B |
> |  GetTimestampUnsafe |                     1 |  1.527 ns | 0.0137 ns | 0.0122 ns | 654,922,326.8 |      - |     - |     - |         - |
> | GetTimestampMarshal |                     6 |  9.355 ns | 0.0174 ns | 0.0136 ns | 106,898,826.1 |      - |     - |     - |         - |
> |  GetTimestampCached |                     6 | 27.531 ns | 0.5606 ns | 0.5757 ns |  36,323,272.1 | 0.0010 |     - |     - |     264 B |
> |  GetTimestampUnsafe |                     6 |  7.831 ns | 0.0193 ns | 0.0150 ns | 127,692,079.6 |      - |     - |     - |         - |
> | GetTimestampMarshal |                    18 | 27.766 ns | 0.0366 ns | 0.0305 ns |  36,015,639.4 |      - |     - |     - |         - |
> |  GetTimestampCached |                    18 | 31.654 ns | 0.3334 ns | 0.2955 ns |  31,591,629.2 | 0.0010 |     - |     - |     264 B |
> |  GetTimestampUnsafe |                    18 | 22.940 ns | 0.0480 ns | 0.0426 ns |  43,592,182.5 |      - |     - |     - |         - |
> | GetTimestampMarshal |                    30 | 54.865 ns | 0.1422 ns | 0.1261 ns |  18,226,681.1 |      - |     - |     - |         - |
> |  GetTimestampCached |                    30 | 45.322 ns | 0.4098 ns | 0.3422 ns |  22,064,253.0 | 0.0010 |     - |     - |     264 B |
> |  GetTimestampUnsafe |                    30 | 38.284 ns | 0.0732 ns | 0.0649 ns |  26,120,299.6 |      - |     - |     - |         - |

```csharp
public class RequestTiming
{
    private readonly ReadOnlyMemory<byte> _data;
    private readonly long[] _values = new long[30];

    public RequestTiming()
    {
        var values = Enumerable.Range(1, 30).Select(i => long.MaxValue - (i * 100)).ToArray();
        _data = new ReadOnlyMemory<byte>(MemoryMarshal.AsBytes(values.AsSpan()).ToArray());
    }

    [Params(1, 6, 18, 30)]
    public int GetTimestampCallCount { get; set; }

    [Benchmark]
    public long[] GetTimestampMarshal()
    {
        for (int i = 0; i < GetTimestampCallCount; i++)
        {
            var valuesSpan = MemoryMarshal.Cast<byte, long>(_data.Span);
            _values[i] = valuesSpan[i];
        }

        return _values;
    }

    [Benchmark]
    public long[] GetTimestampCached()
    {
        long[] valuesArray = MemoryMarshal.Cast<byte, long>(_data.Span).ToArray();

        for (int i = 0; i < GetTimestampCallCount; i++)
        {
            _values[i] = valuesArray[i];
        }

        return _values;
    }

    [Benchmark]
    public long[] GetTimestampUnsafe()
    {
        for (int i = 0; i < GetTimestampCallCount; i++)
        {
            var valuesSpan = MemoryMarshal.CreateSpan(
                ref Unsafe.As<byte, long>(ref MemoryMarshal.GetReference(_data.Span)),
                30);
            _values[i] = valuesSpan[i];
        }

        return _values;
    }
}
```

Reply:
Good call out on the List allocation. I did the inline thing to ensure the compiler didn't optimize the call `var valuesSpan = MemoryMarshal.Cast<byte, long>(_data.Span);` outside of the for loop. 

I think the unsafe idea is good. I didn't know about that so thanks for the tip :)